OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,364 @@
+
+			MaybeShowConnectPanel();
+		}
+
+		static string SanitizedName(string dirty)
+		{
+			if (string.IsNullOrEmpty(dirty))
+				return null;
+
+			// TODO: some special chars are allowed as well, but not at every position
+			var clean = new string(dirty.Where(c => char.IsLetterOrDigit(c)).ToArray());
+
+			if (string.IsNullOrEmpty(clean))
+				return null;
+
+			if (char.IsDigit(clean[0]))
+				return SanitizedName(clean.Substring(1));
+
+			// Source: https://tools.ietf.org/html/rfc2812#section-1.2.1
+			if (clean.Length > 9)
+				clean = clean.Substring(0, 9);
+
+			return clean;
+		}
+
+		void MaybeShowConnectPanel()
+		{
+			if (Irc.Client != null && Irc.Client.IsConnected)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Initialize();
+
+				if (Irc.Client.JoinedChannels.Count > 0)
+					channel = Irc.Client.GetChannel(Irc.Client.JoinedChannels[0]);
+
+				SyncNicknamePanel();
+
+				return;
+			}
+
+			if (Game.Settings.Irc.ConnectAutomatically)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				Connect();
+				return;
+			}
+
+			ircContainer.Visible = false;
+			connectBG.Visible = true;
+
+			var connectAutomaticallyCheckBox = connectBG.Get<CheckboxWidget>("CONNECT_AUTOMATICALLY_CHECKBOX");
+			connectAutomaticallyCheckBox.IsChecked = () => Game.Settings.Irc.ConnectAutomatically;
+			connectAutomaticallyCheckBox.OnClick = () => Game.Settings.Irc.ConnectAutomatically ^= true;
+
+			var connectButton = connectBG.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.IsDisabled = () => string.IsNullOrEmpty(nicknameBox.Text);
+			connectButton.OnClick = () =>
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Game.Settings.Irc.ConnectAutomatically = connectAutomaticallyCheckBox.IsChecked();
+				Game.Settings.Save();
+				Connect();
+			};
+		}
+
+		void Initialize()
+		{
+			Irc.Client.OnConnected += OnConnected;
+			Irc.Client.OnError += OnError;
+			Irc.Client.OnRawMessage += OnRawMessage;
+			Irc.Client.OnJoin += OnJoin;
+			Irc.Client.OnChannelActiveSynced += OnChannelActiveSynced;
+			Irc.Client.OnNickChange += OnNickChange;
+			Irc.Client.OnPart += OnPart;
+			Irc.Client.OnQuit += OnQuit;
+			Irc.Client.OnChannelMessage += OnChannelMessage;
+			Irc.Client.OnPong += OnPong;
+		}
+
+		void Connect()
+		{
+			Irc.Client = new IrcClient();
+			Irc.Client.Encoding = System.Text.Encoding.UTF8;
+			Irc.Client.SendDelay = 100;
+			Irc.Client.ActiveChannelSyncing = true;
+
+			Initialize();
+
+			Game.OnQuit += Disconnect;
+
+			try
+			{
+				AddChatLine("Connecting to {0}...".F(Game.Settings.Irc.Hostname));
+				Irc.Client.Connect(Game.Settings.Irc.Hostname, Game.Settings.Irc.Port);
+			}
+			catch (Exception e)
+			{
+				AddChatLine("Connection error: {0}".F(e.Message));
+				Log.Write("irc", e.ToString());
+			}
+
+			Irc.ListenThread = new Thread(new ThreadStart(Listen));
+			Irc.ListenThread.Name = "IrcListenThread";
+			Irc.ListenThread.Start();
```

Comment:
You don't use the ```ListenThread``` member any more so you can remove it.

I also checked and the library with return from Listen when you call Disconnect, this means you can simply this whole chunk down to:

```new Thread(Irc.Client.Listen) { Name = "IrcListenThread" }.Start();```